Fix deployment center bugs

bishal-pdMSFT opened the pull request:
Fixes https://github.com/Azure/azure-functions-ux/issues/3973, https://github.com/Azure/azure-functions-ux/issues/3975 and #3985

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/Azure/azure-functions-ux?pullRequest=3979) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/Azure/azure-functions-ux?pullRequest=3979) before we can accept your contribution.<br/><br/>:x: bishal-pdMSFT [sign now](https://cla.dotnetfoundation.org/Azure/azure-functions-ux?pullRequest=3979)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/Azure/azure-functions-ux?pullRequest=3979) it.</sub>

bishal-pdMSFT commented:
@nertim can you please help reviewing this PR.

bishal-pdMSFT commented:
Also, added fix for https://github.com/Azure/azure-functions-ux/issues/3985.
Can somebody please help review this PR?

nertim commented:
Hi Bishal... Apologies for the delay. Will take a look today.

nertim commented on client/src/app/shared/models/constants.ts:
```diff
@@ -505,6 +505,10 @@ export class DeploymentCenterConstants {
   public static readonly queueActionFilter = 16; // "Use"
 
+  // PermissionResult
+  public static readonly vstsPermissionApiUri =
```
There is no need for comment. Can you also please move this up with the other vsts URIs?

bishal-pdMSFT replied:
done

nertim commented on server/Resources/Resources.resx:
```diff
@@ -3812,6 +3812,9 @@ Set to "External URL" to use an API definition that is hosted elsewhere.</value>
     <value>Failed to set up deployment: {0}</value>
   </data>
+  <data name="noPermissionsToCreateApp" xml:space="preserve">
+    <value>You can't register a new application in Azure Active Directory. Learn more: https://go.microsoft.com/fwlink/?linkid=843472. More details: {0}</value>
```
We should mention here the reason is due to lack of permission and still provide the link you have posted.

bishal-pdMSFT replied:
The exact permission error message will come from AAD API and will be appended here as "More Details".

nertim commented on client/src/app/site/deployment-center/deployment-center-setup/step-configure/configure-external/configure-external.component.ts:
```diff
@@ -34,8 +34,11 @@ export class ConfigureExternalComponent {
-  public privateRepo = false;
+  public privateRepo = true;
   constructor(public wizard: DeploymentCenterStateManager, private _translateService: TranslateService) {
+    const val = this.wizard.wizardValues;
```
What is going on here? why not just set the this.wizard.wizardValue.sourceSettings.privateRepo = true? also why not use the privateRepo class variable?

bishal-pdMSFT replied:
this.wizard.wizardValues is a getter which returns a property which is readonly and hence can't be mutated. That's why we need to set this.wizard.wizardValues. To me it looks like same pattern is used all the places where wizardValues is used. One such example is in this file on line no 48.
Please let me know if there is an easier way available to update wizardValues.

nertim commented on client/src/app/site/deployment-center/deployment-center-setup/step-configure/configure-external/configure-external.component.ts:
```diff
@@ -34,8 +34,11 @@ export class ConfigureExternalComponent {
+  public privateRepo = true;
   constructor(public wizard: DeploymentCenterStateManager, private _translateService: TranslateService) {
+    const val = this.wizard.wizardValues;
+    val.sourceSettings.privateRepo = true;
```
can any of the properties be null?

nertim commented on client/src/app/site/deployment-center/deployment-center-setup/wizard-logic/deployment-center-state-manager.ts:
```diff
@@ -387,6 +435,33 @@ export class DeploymentCenterStateManager implements OnDestroy {
+  // 2. Session Id - this changes if user does page refresh or changes login
+  // Choosing 2 as it seems longer-lived and is secure
+  private _storeAadCreatePermission(): void {
+    let storedPermissionItem = {
```
Should this be at the app level or user level? Because I believe with this implementation, if the user switches app the selection will presist.

bishal-pdMSFT replied:
This has to be at user level as AAD permissions are at user level. But I could not find any way to get user info other than token info. And token gets refreshed every 60 mins. Hence I chose session level storage. Even if user changes app/keeps the session alive for more than 60 mins, the storage will be persisted.
I am using LocalStorage API and using sessionId as a key to detect if the storage needs to be invalidated. Another option could have been to use SessionStorage, but I did not find any shared service for that nor I could see anyone else using SessionStorage in this repo. Hence I chose LocalStorage. Please let me know if there is a shared service for SessionStorage.

nertim commented on client/src/app/site/deployment-center/provider-dashboards/vso-Dashboard/vso-dashboard.component.ts:
```diff
@@ -127,6 +127,7 @@ export class VsoDashboardComponent implements OnChanges, OnDestroy {
           this._translateService.instant(PortalResources.disconnectingDeployment),
           this._translateService.instant(PortalResources.disconnectingDeployment)
         )
+        .take(1)
```
why this?

nertim replied:
oh looks like this is fixing - https://github.com/Azure/azure-functions-ux/issues/3975 ?

bishal-pdMSFT replied:
Yes

bishal-pdMSFT commented:
> Hi Bishal... I've added some comments. Can you also either add more description to the bugs themselves and/or to the description of this PR? I'm setting a little confused on why the local storage is solution is needed. Rather I am trying to understand what is being solved here. So just a quick update to the description would help. Thank you :)

@nertim I have added more description in one of the comments. But I will be more explicit here:
We are seeing some failures in our telemetry where user does not have permission to either create SPN or assign role to this SPN. This SPN is created by Azure Pipeline on behalf of user to do application deployment and hence the deployment fails. We want such users to get immediate error that they do not have required permissions. 
Furthermore, we want to cache the permission value. So, if we have already made a check for an user we want to avoid repeated checks. For this purpose, we are using Local storage. Another option (as mentioned in the other comment) is to use Session storage. But I could not find any shared service for that and hence decided against using that.

bishal-pdMSFT commented:
@nertim I have resolved comments. Please take a look again.

bishal-pdMSFT commented:
@nertim I do not permission to merge PR. Can you please merge it.
